pydatastructs/miscellaneous_data_structures/segment_tree.py

```diff
@@ -0,0 +1,212 @@
+__all__ = [
+    'SegmentTree'
+]
+
+class SegmentTree:
+    """
+    Represents the Segment Tree
+
+    Parameters
+    ==========
+
+    inp_list: Initial list of values to build the tree.
+
+    Examples
+    ========
+
+    >>> from segment_tree import SegmentTree
+    
+    >>> t = SegmentTree([3, 4, 8, 10, 1])
+    >>> t.tree
+    [0, 26, 25, 1, 7, 18, 1, 0, 3, 4, 8, 10, 1, 0, 0, 0]
+    >>> t.lazy
+    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+    >>> n = t.get_list_len()
+    
+    >>> t.query(1, 0, n - 1, 2, 4)
+    19
+    >>> t.point_update(3, 29)
+    >>> t.tree
+    [0, 45, 44, 1, 7, 37, 1, 0, 3, 4, 8, 29, 1, 0, 0, 0]
+    >>> t.range_update(1, 0, n - 1, 0, 3, -4)
+    >>> t.tree
+    [0, 29, 28, 1, 7, 37, 1, 0, 3, 4, 8, 29, 1, 0, 0, 0]
+    >>> t.lazy
+    [0, 0, 0, 0, -4, -4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+    >>> t.query(1, 0, n - 1, 1, 3)
+    29
+
+    Note
+    ====
+
+    The tree can be modified accordingly to get the [sum, max, min, gcd, etc.]
+    or any other compatible function or query type.
+
+    References
+    ==========
+
+    https://cp-algorithms.com/data_structures/segment_tree.html
+    """
+    def __init__(self, inp_list):
+
+        def get_len(size):
+            """
+            Returns the new list size (the closest power of 2).
+
+            Parameters
+            ==========
+
+            size: Size of the initial list.
+            """
+            if size and size & (size - 1) == 0:
+
+                return size
+            else:
+
+                bit_len = len(bin(size)) - 2
+                return 1 << bit_len
+
+        self.tree_size = 2 * get_len(len(inp_list))
+
+        self.tree = [0 for _ in range(self.tree_size)]
+        self.lazy = [0 for _ in range(self.tree_size)]
+
+        def build(init_list):
+            """
+            Builds the tree using the initial list.
+
+            Parameters
+            ==========
+
+            init_list: Initial list passed in the object declaration.
+            """
+            list_size = len(init_list)
+
+            for i in range(list_size):
+
+                self.tree[self.tree_size // 2 + i] = init_list[i]
+
+            for i in range(list_size, self.tree_size // 2):
+
+                self.tree[self.tree_size // 2 + i] = 0
+
+            for i in range(self.tree_size // 2 - 1, 0, -1):
+
+                self.tree[i] = self.tree[2 * i] + self.tree[2 * i + 1]
```

Comment:
This is just for additions right? How about taking the custom merge functions with 2 arguments while initialising the tree?.
Like: 
`f = lambda a, b: a + b`

Reply:
Yes, I had known about it..will add it..